sem: fix type inference for `static` parameters

zerbina opened the pull request:
## Summary

* fix compiler crash when passing empty aggregate values to `static`
  parameters
* fix `static T` parameter values not being of type `T`, when the
  argument is not directly of type `T` but requires a conversion
* fix converters not being considered for arguments to `static`
  parameters

## Details

When matching an expression against a formal parameter that:
* is a `static` parameter
* contains a `static` type somewhere (e.g., `int or static[float]`)

the expression was compile-time evaluated first, and on success,
assigned a `tyStatic` type. If the `tyStatic` argument type matched the
formal `tyStatic` type, the argument type was bound to the formal
`tyStatic` as-is.

In case an implicit conversion is necessary, the conversion was applied
only *after* the evaluated `tyStatic` was bound to the parameter type,
leading to:
* the actual value inlined at usages of the `static` parameter having
  the wrong type
* type inference for empty aggregate types not working

### Solution

If the formal type is a `static T`, full argument matching (including
implicit conversion and converter handling) for between the argument
and `T` is performed first, and only in case of success is the
expression compile-time evaluated, and the resulting `tyStatic` type
bound to the parameter type.

Post-match fitting has to take place prior to const evaluation (in
order to correctly type empty aggregates), so a new procedure
`tryEvalStaticArgument` is introduced that handles the static
arguments.

The new `static` handling renders the macro/template `static` special-
casing obsolete; replacing the `static` argument with the evaluation
result is now done in `evalTemplateArgs`.

### Typed AST

For calls to routines with `static` arguments, the typed argument
expression now stays in the AST as-is, instead of being replaced with
the evaluated value.

zerbina committed:
sigmatch: handle implicit conversions before evaluation

The evaluated expression now contains conversions, fixing typing
problems with empty container expressions and also making sure
converters are taken into account.

zerbina committed:
sigmatch: move `static` inlining to `evalTemplateArgs`

This is necessary for the original expression to be replaced with their
evaluated-to value. The remaining macro/template static inlining in
`sigmatch` thus becomes obsolete.

zerbina committed:
sigmatch: apply post-match fitting before static evaluation

Without it, empty container types wouldn't be patched up properly.
Instead of doing this in `sigmatch`, the new `tryEvalStaticArgument`
procedure is added to `sem` and made available to `sigmatch` via
`PContext`.

It applies the post-match fitting plus the `tyStatic` wrapping.

zerbina committed:
sigmatch: remove macro/template implicit conversion handling

It's now taken care of in a unified manner by the new `tyStatic` late
evaluation.

zerbina committed:
tests: add test for converter in conjunction with `static`

The test would have previously failed, as `sigmatch` previously looked
for a `float` -> `static int` converter.

zerbina committed:
tests: add more empty container argument test cases

The new ones cover `static` parameters.

zerbina commented:
The failure of `tgeneric_various.nim` is due to proc symbols passed to `static[proc()]` parameters now being turned into closures (as they should). The evaluated result is an `nkClosure`, however, which can currently not pass through `sem` again.

saem commented on compiler/sem/sem.nim:
```diff
@@ -694,6 +694,30 @@ proc semRealConstExpr(c: PContext, n: PNode): PNode =
+    let typ = newTypeS(tyStatic, c)
+    typ.sons = @[result.typ]
+    typ.n = result
+    result = copyNodeWithKids(n)
```
since we already made a copy earlier, do we need to make a copy again?

zerbina replied:
Yep, we do, as `prepareNode` would run into infinite recursion otherwise. Consider the following situation:
1. `n` is passed to `tryConstExpr`
2. if `n` is already a constant expression and there's nothing to fold or evaluate, `tryConstExpr` returns `n` as-is
3. `result == n` is true in that case
4. a `tyStatic` with `PType.n` set to `result` (i.e., `n`) is created

If `result` would be set to `n` (without a copy), the `result.typ = typ` assignment below would result in a type where `typ.n.typ == typ`.

The shallow copy makes sure that `typ.n.typ != typ`, though I could change `result = copyNodeWithKids(n)` to:
```nim
result =
  if n != result: n
  else:           copyNodeWithKids(n)
```
in order to get around an unnecessary shallow copy, and to also emphasize why the copy is needed.

saem replied:
Thanks for the details explanation, I forgot that `tryConstExpr` can act like an identity function in certain cases. I agree, the suggested code that avoids the copy is clearer in that regard.

zerbina replied:
I've adjusted the code a bit -- it should be clearer now.

zerbina committed:
sem: refactor `tryEvalStaticArgument`

Restructure the type creation a bit, with the goal of making it easier
to follow.

zerbina committed:
semexprs: handle `nkClosure` in `semExpr`

Only the no-environment form is handled, as closures with an
environment shouldn't be able to leave compile-time evaluation in the
first place.

zerbina committed:
semexprs: ensure proper typing for `nkClosure` proc

Use `fitNode` to restore the implicit conversion, which makes sure that
an error is reported in case there's a type mismatch (only possible in
macro-generated AST).

zerbina commented:
For making `static[proc()]` parameters work again, I've added handling for `nkClosure` to `semExpr` -- `nkClosure` AST is now turned back into an `nkHiddenStdConv`. I think it's an okay solution for now; the alternatives would have been to either:
* implement full analysis for `nkClosure`
* OR turn closures without an environment into just an `nkSym` node in `vmcompilerserdes`

saem reviewed (approved):
I think the this solution, where `semexprs` handles `nkClosure` is better as it's less action at a distance (compared to `vmserdes`), and full `nkClosure` analysis would be built into `sem`, likely involving `semexprs` anyhow.

saem commented:
/merge
